go/vt/workflow/manager.go

```diff
@@ -0,0 +1,374 @@
+	_, err := m.ts.CreateWorkflow(ctx, w)
+	if err != nil {
+		return "", err
+	}
+
+	// And we're done.
+	return w.Uuid, nil
+}
+
+// Start with start a Workflow. It will load it in memory, update its
+// status to Running, and call its Run() method.
+func (m *Manager) Start(ctx context.Context, uuid string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Check the manager is running.
+	if m.ctx == nil {
+		return fmt.Errorf("manager not running")
+	}
+
+	// Make sure it is not running already.
+	if _, ok := m.workflows[uuid]; ok {
+		return fmt.Errorf("workflow %v is already running", uuid)
+	}
+
+	// Load it from the topo server, make sure it has the right state.
+	wi, err := m.ts.GetWorkflow(ctx, uuid)
+	if err != nil {
+		return err
+	}
+	if wi.State != workflowpb.WorkflowState_NotStarted {
+		return fmt.Errorf("workflow with uuid %v is in state %v", uuid, wi.State)
+	}
+	factory, ok := factories[wi.FactoryName]
+	if !ok {
+		return fmt.Errorf("workflow %v is using factory name %v but no such factory registered", uuid, wi.FactoryName)
+	}
+
+	// Change its state in the topo server. Note we do that first,
+	// so if the running part fails, we will retry next time.
+	wi.State = workflowpb.WorkflowState_Running
+	wi.StartTime = time.Now().Unix()
+	if err := m.ts.SaveWorkflow(ctx, wi); err != nil {
+		return err
+	}
+
+	// Ask the factory to create the running workflow.
+	w, err := factory.Instantiate(wi.Workflow)
+	if err != nil {
+		return err
+	}
+
+	// Create a context to run it, and save the runningWorkflow.
+	ctx, cancel := context.WithCancel(m.ctx)
+	rw := &runningWorkflow{
+		wi:       wi,
+		cancel:   cancel,
+		workflow: w,
+		done:     make(chan struct{}),
+	}
+	m.workflows[uuid] = rw
+
+	// And run it in the background.
+	go m.runWorkflow(ctx, uuid, rw)
+
+	return nil
+}
+
+func (m *Manager) runWorkflow(ctx context.Context, uuid string, rw *runningWorkflow) {
+	defer close(rw.done)
+
+	err := rw.workflow.Run(ctx, m, rw.wi)
+
+	// Change the Manager state.
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.ctx == nil {
+		// We are shutting down, nothing else to do.
+		return
+	}
+
+	// We are not shutting down, but this workflow is done, or
+	// canceled. In any case, delete it from the running
+	// workflows, and change its topo Server state.
+	delete(m.workflows, uuid)
+	rw.wi.State = workflowpb.WorkflowState_Done
+	if err != nil {
+		rw.wi.Error = err.Error()
+	}
+	rw.wi.EndTime = time.Now().Unix()
+	if err := m.ts.SaveWorkflow(m.ctx, rw.wi); err != nil {
+		log.Errorf("Could not save workflow %v after completion: %v", rw.wi, err)
+	}
+}
+
+// Stop stops the running workflow. It will cancel its context and
+// wait for it to exit.
+func (m *Manager) Stop(ctx context.Context, uuid string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Find the workflow.
+	rw, ok := m.workflows[uuid]
+	if !ok {
+		return fmt.Errorf("no running workflow with uuid %v", uuid)
+	}
+
+	// Cancel the running guy, and waits for it.
+	rw.cancel()
```

Comment:
You're right it probably doesn't. I'll release the lock before waiting. I'll cover this in unit tests.